src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs

```diff
@@ -66,155 +73,195 @@ public static int CacheSize
+
+            CultureInfo culture = (options & RegexOptions.CultureInvariant) != 0 ? CultureInfo.InvariantCulture : CultureInfo.CurrentCulture;
+            Key key = new Key(pattern, culture.ToString(), options, matchTimeout != Regex.InfiniteMatchTimeout);
+
+            if (!TryGet(key, out Regex? regex))
             {
-                // first look for it in the cache and move it to the head
-                CachedCodeEntry? entry = LookupCachedAndPromote(key);
+                regex = new Regex(pattern, options, matchTimeout, culture);
+                Add(key, regex);
+            }
 
-                // it wasn't in the cache, so we'll add a new one
-                if (entry == null && isToAdd && s_cacheSize != 0) // check cache size again in case it changed
+            return regex;
+        }
+
+        private static bool TryGet(Key key, [NotNullWhen(true)] out Regex? regex)
+        {
+            Node? cachedRegex = s_cacheFirst;
+            if (cachedRegex != null)
+            {
+                if (cachedRegex.Key.Equals(key))
                 {
-                    entry = new CachedCodeEntry(key, capnames!, capslist!, _code!, caps!, capsize, _runnerref!, _replref!);
+                    regex = cachedRegex.Regex;
+                    return true;
+                }
 
-                    // put first in linked list:
-                    if (s_cacheFirst != null)
+                if (s_maxCacheSize != 0)
+                {
+                    lock (s_cache)
                     {
-                        SysDebug.Assert(s_cacheFirst.Next == null);
-                        s_cacheFirst.Next = entry;
-                        entry.Previous = s_cacheFirst;
+                        cachedRegex = LookupCachedAndPromote(key);
+                        if (cachedRegex != null)
+                        {
+                            regex = cachedRegex.Regex;
+                            return true;
+                        }
                     }
-                    s_cacheFirst = entry;
+                }
+            }
 
-                    s_cacheCount++;
-                    if (s_cacheCount >= CacheDictionarySwitchLimit)
+            regex = null;
+            return false;
+        }
+
+        private static void Add(Key key, Regex regex)
+        {
+            lock (s_cache)
+            {
+                // If we're not supposed to cache, or if the entry is already cached, we're done.
+                if (s_maxCacheSize == 0 || LookupCachedAndPromote(key) != null)
+                {
+                    return;
+                }
+
+                // Create the entry for caching.
+                var entry = new Node(key, regex);
+
+                // Put it at the beginning of the linked list, as it is the most-recently used.
+                if (s_cacheFirst != null)
+                {
+                    Debug.Assert(s_cacheFirst.Next == null);
+                    s_cacheFirst.Next = entry;
```

Comment:
I find myself confused by the nomenclature here. I know this is how the current code is - and so maybe this just me thinking about it incorrectly.

But if I have a list of things, let's say pages in a book, and I have the "first" page and the "last" page. The way I would expect to get from the "first" to the "last" is to keep going to the "next" page until I've hit the last. And the way I would go from "last" to "first" is to keep going to the "previous" page until I've hit the first.

The way I understand this code, is that it is flipped from my mental model. I was wondering how you thought about it.

Reply:
You can kind of see this confusion in the `for` loop below:

```C#
for (Node? next = s_cacheFirst; next != null; next = next.Previous)
{
    s_cache.Add(next.Key, next);
}
```

Reply:
My next PR is going to delete all of this, so I'm going to leave it as-is for now ;)

Reply:
oh? well that sounds good to me then.